File: app/src/main/java/io/treehouses/remote/bases/BaseSSHConsole.java
package io.treehouses.remote.bases;

import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.widget.Toolbar;
import androidx.viewpager.widget.PagerAdapter;

import com.google.android.material.tabs.TabLayout;

import java.util.ArrayList;

import io.treehouses.remote.R;
import io.treehouses.remote.SSH.PromptHelper;
import io.treehouses.remote.SSH.Terminal.TerminalBridge;
import io.treehouses.remote.SSH.Terminal.TerminalManager;
import io.treehouses.remote.SSH.Terminal.TerminalView;
import io.treehouses.remote.SSH.Terminal.TerminalViewPager;
import io.treehouses.remote.SSH.beans.HostBean;

public class BaseSSHConsole extends BaseFragment {
//    protected TerminalManager bound = null;
//    protected TerminalPagerAdapter adapter;
//    protected RelativeLayout stringPromptGroup;
//    protected EditText stringPrompt;
//    private RelativeLayout booleanPromptGroup;
//    private TextView booleanPrompt;
//    private boolean keyboardAlwaysVisible = false;
//    private Button booleanYes;
//    private LinearLayout keyboardGroup;
//    private Runnable keyboardGroupHider;
//    protected TerminalViewPager pager = null;
//    protected TabLayout tabs = null;
//    protected Toolbar toolbar = null;
//    private Handler handler = new Handler();
//
//    @Nullable private ActionBar actionBar;
//    private boolean inActionBarMenu = false;
//    private boolean titleBarHide;
//
//    private Animation fade_out_delayed;
//
//
//
//    private TextView stringPromptInstructions;
////    protected TerminalPagerAdapter adapter = null;
//    protected LayoutInflater inflater = null;
//
//    @Nullable
//    @Override
//    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
//        fade_out_delayed = AnimationUtils.loadAnimation(getContext(), R.anim.fade_out_delayed);
//        return super.onCreateView(inflater, container, savedInstanceState);
//    }
//
//    protected void hideAllPrompts() {
//        stringPromptGroup.setVisibility(View.GONE);
//        booleanPromptGroup.setVisibility(View.GONE);
//    }
//    protected void updatePromptVisible() {
//        // check if our currently-visible terminalbridge is requesting any prompt services
//        TerminalView view = adapter.getCurrentTerminalView();
//
//        // Hide all the prompts in case a prompt request was canceled
//        hideAllPrompts();
//
//        if (view == null) {
//            // we dont have an active view, so hide any prompts
//            return;
//        }
//
//        PromptHelper prompt = view.bridge.promptHelper;
//        if (String.class.equals(prompt.promptRequested)) {
//            hideEmulatedKeys();
//            stringPromptGroup.setVisibility(View.VISIBLE);
//
//            String instructions = prompt.promptInstructions;
//            if (instructions != null && instructions.length() > 0) {
//                stringPromptInstructions.setVisibility(View.VISIBLE);
//                stringPromptInstructions.setText(instructions);
//            } else
//                stringPromptInstructions.setVisibility(View.GONE);
//            stringPrompt.setText("");
//            stringPrompt.setHint(prompt.promptHint);
//            stringPrompt.requestFocus();
//
//        } else if (Boolean.class.equals(prompt.promptRequested)) {
//            hideEmulatedKeys();
//            booleanPromptGroup.setVisibility(View.VISIBLE);
//            booleanPrompt.setText(prompt.promptHint);
//            booleanYes.requestFocus();
//
//        } else {
//            hideAllPrompts();
//            view.requestFocus();
//        }
//    }
//    private void hideEmulatedKeys() {
//        if (!keyboardAlwaysVisible) {
//            if (keyboardGroupHider != null)
//                handler.removeCallbacks(keyboardGroupHider);
//            keyboardGroup.setVisibility(View.GONE);
//        }
//        hideActionBarIfRequested();
//    }
//    private void hideActionBarIfRequested() {
//        if (titleBarHide && actionBar != null) {
//            actionBar.hide();
//        }
//    }
//    protected Handler promptHandler = new Handler() {
//        @Override
//        public void handleMessage(Message msg) {
//            // someone below us requested to display a prompt
//            updatePromptVisible();
//        }
//    };
//
//    public class TerminalPagerAdapter extends PagerAdapter {
//        @Override
//        public int getCount() {
//            if (bound != null) {
//                return bound.bridges.size();
//            } else {
//                return 0;
//            }
//        }
//
//        @Override
//        public Object instantiateItem(ViewGroup container, int position) {
//            if (bound == null || bound.bridges.size() <= position) {
//                Log.w("TAG", "Activity not bound when creating TerminalView.");
//            }
//            TerminalBridge bridge = bound.bridges.get(position);
//            bridge.promptHelper.setHandler(promptHandler);
//
//            // inflate each terminal view
//            RelativeLayout view = (RelativeLayout) inflater.inflate(R.layout.item_terminal, container, false);
//
//            // set the terminal name overlay text
//            TextView terminalNameOverlay = view.findViewById(R.id.terminal_name_overlay);
//            terminalNameOverlay.setText(bridge.host.getNickname());
//
//            // and add our terminal view control, using index to place behind overlay
//            final TerminalView terminal = new TerminalView(container.getContext(), bridge, pager);
//            terminal.setId(R.id.terminal_view);
//            view.addView(terminal, 0);
//
//            // Tag the view with its bridge so it can be retrieved later.
//            view.setTag(bridge);
//
//            container.addView(view);
//            terminalNameOverlay.startAnimation(fade_out_delayed);
//            return view;
//        }
//
//        @Override
//        public void destroyItem(ViewGroup container, int position, Object object) {
//            final View view = (View) object;
//
//            container.removeView(view);
//        }
//
//        @Override
//        public int getItemPosition(Object object) {
//            if (bound == null) {
//                return POSITION_NONE;
//            }
//
//            View view = (View) object;
//            TerminalView terminal = view.findViewById(R.id.terminal_view);
//            HostBean host = terminal.bridge.host;
//
//            int itemIndex = POSITION_NONE;
//            int i = 0;
//            for (TerminalBridge bridge : bound.bridges) {
//                if (bridge.host.equals(host)) {
//                    itemIndex = i;
//                    break;
//                }
//                i++;
//            }
//            return itemIndex;
//        }
//
//        public TerminalBridge getBridgeAtPosition(int position) {
//            if (bound == null) {
//                return null;
//            }
//
//            ArrayList<TerminalBridge> bridges = bound.bridges;
//            if (position < 0 || position >= bridges.size()) {
//                return null;
//            }
//            return bridges.get(position);
//        }
//
//        @Override
//        public void notifyDataSetChanged() {
//            super.notifyDataSetChanged();
//            if (tabs != null) {
//                toolbar.setVisibility(this.getCount() > 1 ? View.VISIBLE : View.GONE);
//                tabs.setTabsFromPagerAdapter(this);
//            }
//        }
//
//        @Override
//        public boolean isViewFromObject(View view, Object object) {
//            return view == object;
//        }
//
//        @Override
//        public CharSequence getPageTitle(int position) {
//            TerminalBridge bridge = getBridgeAtPosition(position);
//            if (bridge == null) {
//                return "???";
//            }
//            return bridge.host.getNickname();
//        }
//
//        public TerminalView getCurrentTerminalView() {
//            View currentView = pager.findViewWithTag(getBridgeAtPosition(pager.getCurrentItem()));
//            if (currentView == null) {
//                return null;
//            }
//            return (TerminalView) currentView.findViewById(R.id.terminal_view);
//        }
//    }
}

File: app/src/main/kotlin/io/treehouses/remote/BaseSSHConsole.kt
package io.treehouses.remote

import android.annotation.TargetApi
import android.content.pm.ActivityInfo
import android.content.res.Configuration
import android.os.Bundle
import android.os.Handler
import android.util.Log
import android.view.MotionEvent
import android.view.View
import android.view.Window
import android.view.animation.Animation
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.lifecycle.ViewModelProvider
import io.treehouses.remote.SSH.PromptHelper
import io.treehouses.remote.SSH.Terminal.TerminalKeyListener
import io.treehouses.remote.SSH.Terminal.TerminalView
import io.treehouses.remote.databinding.ActivitySshConsoleBinding

open class BaseSSHConsole: AppCompatActivity() {

    protected lateinit var bind: ActivitySshConsoleBinding
    lateinit var viewModel: SSHConsoleViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewModel = ViewModelProvider(this)[SSHConsoleViewModel::class.java]!!
    }

    @TargetApi(11)
    protected fun requestActionBar() {
        supportRequestWindowFeature(Window.FEATURE_ACTION_BAR_OVERLAY)
    }

    protected fun updatePrompt(b:Boolean, helper: PromptHelper) {
        helper.setResponse(b)
        updatePromptVisible()
    }

    /**
     * Show any prompts requested by the currently visible [TerminalView].
     */
    protected fun updatePromptVisible() {
        // check if our currently-visible terminalbridge is requesting any prompt services
        val view = viewModel.adapter?.currentTerminalView

        // Hide all the prompts in case a prompt request was canceled
        hideAllPrompts()
        // we dont have an active view, so hide any prompts
        if (view == null) return
        val prompt = view.bridge.promptHelper
        Log.e("GOT", "HERE ")
        when {
            String::class.java == prompt!!.promptRequested -> {
                hideEmulatedKeys()
                setConsolePassword(prompt)
            }
            Boolean::class.java == prompt.promptRequested -> {
                hideEmulatedKeys()
                bind.consoleBooleanGroup.visibility = View.VISIBLE
                bind.consolePrompt.text = prompt.promptHint
                bind.consolePromptYes.requestFocus()
            }
            else -> {
                hideAllPrompts()
                view.requestFocus()
            }
        }
    }

    protected fun hideEmulatedKeys() {
        if (!viewModel.keyboardAlwaysVisible) {
            if (viewModel.keyboardGroupHider != null) viewModel.handler.removeCallbacks(viewModel.keyboardGroupHider!!)
            bind.keyboard.keyboardGroup.visibility = View.GONE
        }
        hideActionBarIfRequested()
    }

    private fun hideActionBarIfRequested() {
        if (viewModel.titleBarHide && viewModel.actionBar != null) viewModel.actionBar!!.hide()
    }

    private fun setConsolePassword(prompt: PromptHelper) {
        bind.consolePasswordGroup.visibility = View.VISIBLE
        val instructions = prompt.promptInstructions
        if (instructions != null && instructions.isNotEmpty()) {
            bind.consolePasswordInstructions.visibility = View.VISIBLE
            bind.consolePasswordInstructions.text = instructions
        } else bind.consolePasswordInstructions.visibility = View.GONE
        bind.consolePassword.setText("")
        bind.consolePassword.hint = prompt.promptHint
        bind.consolePassword.requestFocus()
    }

    private fun hideAllPrompts() {
        bind.consolePasswordGroup.visibility = View.GONE
        bind.consoleBooleanGroup.visibility = View.GONE
    }

    protected fun updateEmptyVisible() {
        // update visibility of empty status message
        viewModel.empty!!.visibility = if (bind.pager.childCount == 0) View.VISIBLE else View.GONE
    }

    /**
     * @param bridge
     */
    protected fun closeBridge() {
        updateEmptyVisible()
        updatePromptVisible()
        // If we just closed the last bridge, go back to the previous activity.
        if (bind.pager.childCount == 0) {
            Log.e("FINISHING SSH", "FINISH")
            finish()
        }
    }

    protected fun onEmulatedKeyClicked(v: View) {
        val terminal = viewModel.adapter!!.currentTerminalView ?: return
        val handler = terminal.bridge.keyHandler
        var hideKeys = sendKeys(v, handler)
        if (hideKeys) hideEmulatedKeys()
        else autoHideEmulatedKeys()

        terminal.bridge.tryKeyVibrate()
        hideActionBarIfRequested()
    }

    private fun sendKeys(v: View, handler: TerminalKeyListener) : Boolean {
        return if (viewModel.isSpecialButton(v, handler)) true
        else {
            viewModel.checkButtons(v, handler)
            false
        }
    }

    protected fun autoHideEmulatedKeys() {
        if (viewModel.keyboardGroupHider != null) viewModel.handler.removeCallbacks(viewModel.keyboardGroupHider!!)
        viewModel.keyboardGroupHider = Runnable {
            if (bind.keyboard.keyboardGroup.visibility == View.GONE || viewModel.inActionBarMenu) {
                return@Runnable
            }
            if (!viewModel.keyboardAlwaysVisible) setAnimation(viewModel.keyboardFadeOut, View.GONE)
            hideActionBarIfRequested()
            viewModel.keyboardGroupHider = null
        }
        viewModel.handler.postDelayed(viewModel.keyboardGroupHider!!, KEYBOARD_DISPLAY_TIME.toLong())
    }

    protected fun setAnimation(animation: Animation?, visibility: Int) {
        bind.keyboard.keyboardGroup.startAnimation(animation)
        bind.keyboard.keyboardGroup.visibility = visibility
    }

    protected fun checkSession(view: TerminalView?, activeTerminal: Boolean): Pair<Boolean, Boolean> {
        var sessionOpen = false
        var disconnected = false
        if (activeTerminal) {
            val bridge = view!!.bridge
            sessionOpen = bridge.isSessionOpen
            disconnected = bridge.isDisconnected
        }
        return Pair(sessionOpen, disconnected)
    }

    /**
     * Handle repeatable virtual keys and touch events
     */
    inner class KeyRepeater(private val mHandler: Handler, private val mView: View) : Runnable, View.OnTouchListener, View.OnClickListener {
        private var mDown = false
        override fun run() {
            mDown = true
            mHandler.removeCallbacks(this)
            mHandler.postDelayed(this, KEYBOARD_REPEAT.toLong())
            mView.performClick()
        }

        override fun onTouch(v: View, event: MotionEvent): Boolean {
            when (event.action) {
                MotionEvent.ACTION_DOWN -> {
                    mDown = false
                    mHandler.postDelayed(this, KEYBOARD_REPEAT_INITIAL.toLong())
                    mView.isPressed = true
                }
                MotionEvent.ACTION_CANCEL -> {
                    mHandler.removeCallbacks(this)
                    mView.isPressed = false
                }
                MotionEvent.ACTION_UP -> {
                    mHandler.removeCallbacks(this)
                    mView.isPressed = false
                    if (!mDown) mView.performClick()
                }
                else -> return false
            }
            return true
        }

        override fun onClick(view: View) = onEmulatedKeyClicked(view)

    }

    /**
     *
     */
    protected fun configureOrientation() {
        val rotateDefault: String = if (resources.configuration.keyboard == Configuration.KEYBOARD_NOKEYS) PreferenceConstants.ROTATION_PORTRAIT else PreferenceConstants.ROTATION_LANDSCAPE
        var rotate = viewModel.prefs!!.getString(PreferenceConstants.ROTATION, rotateDefault)
        if (PreferenceConstants.ROTATION_DEFAULT == rotate) rotate = rotateDefault

        // request a forced orientation if requested by user
        when (rotate) {
            PreferenceConstants.ROTATION_LANDSCAPE -> {
                requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE
                viewModel.forcedOrientation = true
            }
            PreferenceConstants.ROTATION_PORTRAIT -> {
                requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
                viewModel.forcedOrientation = true
            }
            else -> {
                requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED
                viewModel.forcedOrientation = false
            }
        }
    }

    protected fun setIntentRequested() {
        val requestedBridge = viewModel.bound!!.mNicknameBridgeMap[viewModel.requested!!.fragment]?.get()
        var requestedIndex = 0
        synchronized(bind.pager) {
            if (requestedBridge == null) {
                // If we didn't find the requested connection, try opening it
                try {
                    Log.d(TAG, String.format("We couldnt find an existing bridge with URI=%s (nickname=%s)," +
                            "so creating one now", viewModel.requested.toString(), viewModel.requested!!.fragment))
                    viewModel.bound!!.openConnection(viewModel.requested)
                } catch (e: Exception) {
                    Log.e(TAG, "Problem while trying to create new requested bridge from URI", e)
                    return
                }
                viewModel.adapter!!.notifyDataSetChanged()
                requestedIndex = viewModel.adapter!!.count
            } else {
                val flipIndex = viewModel.bound!!.bridges.indexOf(requestedBridge)
                if (flipIndex > requestedIndex) {
                    requestedIndex = flipIndex
                }
            }
            setDisplayedTerminal(requestedIndex)
        }
    }

    /**
     * Displays the child in the ViewPager at the requestedIndex and updates the prompts.
     *
     * @param requestedIndex the index of the terminal view to display
     */
    protected fun setDisplayedTerminal(requestedIndex: Int) {
        bind.pager.currentItem = requestedIndex
        // set activity title
        title = viewModel.adapter!!.getPageTitle(requestedIndex)
        onTerminalChanged()
    }

    private fun findCurrentView(id: Int): View? {
        val view = bind.pager.findViewWithTag<View>(viewModel.adapter!!.getBridgeAtPosition(bind.pager.currentItem))
                ?: return null
        return view.findViewById(id)
    }

    /**
     * Called whenever the displayed terminal is changed.
     */
    protected fun onTerminalChanged() {
        val terminalNameOverlay = findCurrentView(R.id.terminal_name_overlay)
        terminalNameOverlay?.startAnimation(viewModel.fadeOutDelayed)
        viewModel.updateDefault()
        updatePromptVisible()
        ActivityCompat.invalidateOptionsMenu(this@BaseSSHConsole)
    }

    companion object {
        const val TAG = "CB.ConsoleActivity"
        const val KEYBOARD_DISPLAY_TIME = 3000
        const val KEYBOARD_REPEAT_INITIAL = 500
        const val KEYBOARD_REPEAT = 100
        const val STATE_SELECTED_URI = "selectedUri"
    }
}